For E0223, suggest associated functions that are similar to the path

trevyn opened the pull request:
e.g. for `String::from::utf8`, suggest `String::from_utf8`

Closes #109195

trevyn committed:
For E0223, suggest methods that look similar to the path

oli-obk commented on compiler/rustc_hir_analysis/src/astconv/mod.rs:
```diff
@@ -1373,6 +1374,57 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
+                            {
+                                true
+                            } else {
+                                qself_ty.is_str()
+                                    && ["from_utf8", "from_utf8_mut"].contains(&name.as_str())
```
imo this is too special cased for us to include here. Also, why is this not achieved with the above logic?

oli-obk commented on compiler/rustc_hir_analysis/src/astconv/mod.rs:
```diff
@@ -1373,6 +1374,57 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
+                        }) = node
+                        {
+                            let name = format!("{ident2}_{ident3}");
+                            if if let Some(ty_def_id) = qself_ty.ty_def_id()
+                                && let Ok([inherent_impl]) = tcx.inherent_impls(ty_def_id)
+                                && let Some(ty::AssocItem { kind: ty::AssocKind::Fn, .. }) = tcx
+                                    .associated_items(inherent_impl)
+                                    .filter_by_name_unhygienic(Symbol::intern(&name))
+                                    .next()
```
Maybe use something like the [method resolution logic](https://github.com/rust-lang/rust/blob/master/compiler/rustc_hir_typeck/src/callee.rs#L87) to find the method, even if it's hidden behind derefs.

trevyn replied:
https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/Method.20resolution.20logic.20from.20.60AstConv.60

trevyn committed:
Create helper `maybe_report_similar_assoc_fn`

trevyn commented:
@oli-obk I moved everything into a helper function and removed the `str` special-casing. I don't quite follow how to get to method lookup from `rustc_hir_analysis`/`AstConv`, a pointer would be helpful if this still makes sense to you.

oli-obk commented:
Oh... yea, there's no good way to invoke that from here :/ I guess we'll miss out on some cases, but get the usual ones.

@bors r+ rollup

bors commented:
:pushpin: Commit 0b6af718d85a2639854ad09730f5f39aa5f70f9a has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 0b6af718d85a2639854ad09730f5f39aa5f70f9a -->
<!-- homu: {"type":"Approved","sha":"0b6af718d85a2639854ad09730f5f39aa5f70f9a","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
